src/content/effective-dart/design.md

```diff
@@ -1410,47 +1410,37 @@ var completer = Completer<Map<String, int>>();
 ```
 
 
-### DO annotate with `dynamic` instead of letting inference fail
+<a id="do-annotate-with-dynamic-instead-of-letting-inference-fail" aria-hidden="true"></a>
+### DO annotate with `Object?` instead of letting inference fail
 
-When inference doesn't fill in a type, it usually defaults to `dynamic`. If
-`dynamic` is the type you want, this is technically the most terse way to get
-it. However, it's not the most *clear* way. A casual reader of your code who
-sees that an annotation is missing has no way of knowing if you intended it to be
-`dynamic`, expected inference to fill in some other type, or simply forgot to
-write the annotation.
+When inference doesn't fill in a type, it usually defaults to `dynamic`, which
+is rarely the best type to use. A `dynamic` reference allows for unsafe
+operations that use identical syntax to operations that are statically safe when
+the type is not `dynamic`. An `Object?` reference is safer. For example a
+`dynamic` reference may fail a type cast that is not visible in the syntax,
+while an `Object?` reference will guarantee that the `as` cast is written in
+source.
 
-When `dynamic` is the type you want, write that explicitly to make your intent
-clear and highlight that this code has less static safety.
+Use `Object?` to indicate in a signature that any type of object, or null, is
+allowed.
```

Comment:
The overall metarule for the guidelines is "unless stated otherwise, express things the most terse way". If you want some return type or parameter type to have type `dynamic`, the most terse way is to simply write nothing.

The original intent of this guideline was to opt *out* of that default "use the shortest thing" rule. That's why the guideline is about "instead of letting inference fail".

The new guideline here is really about "when should you use `Object?` versus `dynamic`, which is a different thing. I think the second updated guideline below does a pretty good job of covering that.

So for this one, I think it would be better to recast it to something like `DON'T let inference fall back to dynamic".